### ISSUE TITLE
System.Net.Security.SslStream throws exception while running EF tests

### ISSUE THREAD

**Issue by saurabh500** [dotnet/corefx#27006]
While running EF tests on Ubuntu against netcoreapp2.1 an exception is thrown 

```
Failed   Microsoft.EntityFrameworkCore.SqlServerDatabaseCreatorEnsureDeletedTest.Deletes_database(async: False, open: True, ambientTransaction: False)
Error Message:
 System.NotSupportedException :  The WriteAsync method cannot be called when another write operation is pending.
Stack Trace:
   at System.Net.Security.SslStreamInternal.WriteAsyncInternal[TWriteAdapter](TWriteAdapter writeAdapter, ReadOnlyMemory`1 buffer)
   at System.Net.Security.SslStreamInternal.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Security.SslStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.Data.SqlClient.SNI.SNITCPHandle.Send(SNIPacket packet)
   at System.Data.SqlClient.SNI.SNIMarsConnection.Send(SNIPacket packet)
   at System.Data.SqlClient.SNI.SNIMarsHandle.SendControlPacket(SNISMUXFlags flags)
   at System.Data.SqlClient.SNI.SNIMarsHandle.Dispose()
   at System.Data.SqlClient.SNI.TdsParserStateObjectManaged.Dispose()
   at System.Data.SqlClient.TdsParser.Disconnect()
   at System.Data.SqlClient.SqlInternalConnectionTds.Dispose()
   at System.Data.ProviderBase.DbConnectionPool.DestroyObject(DbConnectionInternal obj)
   at System.Data.ProviderBase.DbConnectionPool.DeactivateObject(DbConnectionInternal obj)
   at System.Data.ProviderBase.DbConnectionPool.PutObject(DbConnectionInternal obj, Object owningObject)
   at System.Data.ProviderBase.DbConnectionInternal.CloseConnection(DbConnection owningObject, DbConnectionFactory connectionFactory)
   at System.Data.SqlClient.SqlConnection.Close()
   at System.Data.SqlClient.SqlConnection.Dispose(Boolean disposing)
   at System.ComponentModel.Component.Dispose()
   at Microsoft.EntityFrameworkCore.TestUtilities.RelationalTestStore.Dispose() in /home/saurabh/EntityFrameworkCore/src/EFCore.Relational.Specification.Tests/TestUtilities/RelationalTestStore.cs:line 41
   at Microsoft.EntityFrameworkCore.TestUtilities.SqlServerTestStore.Dispose() in /home/saurabh/EntityFrameworkCore/test/EFCore.SqlServer.FunctionalTests/TestUtilities/SqlServerTestStore.cs:line 414
   at Microsoft.EntityFrameworkCore.SqlServerDatabaseCreatorEnsureDeletedTest.<Delete_database_test>d__2.MoveNext() in /home/saurabh/EntityFrameworkCore/test/EFCore.SqlServer.FunctionalTests/SqlServerDatabaseCreatorTest.cs:line 154
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---
--- End of stack trace from previous location where exception was thrown ---

```

Steps to repro 

Clone the EFCore repo `https://github.com/aspnet/EntityFrameworkCore`

Build the repo from the root using build.sh

Execute the tests in /EntityFrameworkCore/test/EFCore.SqlServer.FunctionalTests
Steps:
1. Modify config.json to have a valid connection string (something like) "Data Source=tcp:server,1437;Database=master;Integrated Security=False;Connect Timeout=60;ConnectRetryCount=0;uid=saurabh;pwd=*******;MultipleActiveResultSets=False"

2. Execute test 
```
 ~/.dotnet/dotnet test --framework netcoreapp2.1 --filter="FullyQualifiedName=Microsoft.EntityFrameworkCore.SqlServerDatabaseCreatorEnsureDeletedTest.Deletes_database"
```

This test passes one out of every 5 times when run for framework netcoreapp2.1 
Run the test against netcoreapp2.0 and it passes consistently. 

Some in System.Net.Security changed? 

Please contact me for the database password via email for quicker setup.

**Comment by smitpatel** [comment]
I haven't seen such an error yet in our CI builds. I will run it locally on Linux & mac both and see if I encounter same issues.

**Comment by stephentoub** [comment]
It's of course possible there's an SslStream bug, though at first glance it sounds higher up and that the error is accurate: some code is trying to make a second write while a previous one is still in flight. That's not supported, nor was it previously. There have been changes to SslStream in 2.1, which could potentially be making a race condition in consuming code more likely to happen.

**Comment by saurabh500** [comment]
@stephentoub thanks. I will see if I can isolate the scenario. It won't be easy to provide a standalone repro in this case with only sslstream and Sql. However I can try to figure out a repro with SqlClient and sslstream. Meanwhile if this bug can be investigated with  the test case above that would be great.

**Comment by saurabh500** [comment]
@karelz why does this bug have a need more info label?

**Comment by karelz** [comment]
@saurabh500 per Stephen's analysis above, it may be very much a problem in the user code (the test), hence we need more clarity if that's the case. If there is evidence it is a bug in SslStream, rather than in calling code, we will be more than happy to investigate for 2.1.

**Comment by saurabh500** [comment]
The same code works fine for netcoreapp2.0 and fails in netcoreapp2.1
I think that should be enough information to consider this as a regression. 
In case it is figured that SqlClient had been calling sslstream wrong from the very beginning I would be happy to change it. 
I think there is enough information to investigate in this bug.

**Comment by karelz** [comment]
The 20 vs. 21 difference might be from this "_There have been changes to SslStream in 2.1, which could potentially be making a race condition in consuming code more likely to happen._".
Given that information I don't consider it a regression in CoreFX yet - therefore Future + needs more information.

It would be much easier if someone familiar with the test can take it out of the test harness to simplify the repro and isolate it for easier debugging.
From brief look at [the test](https://github.com/aspnet/EntityFrameworkCore/blob/f6bb8c19a37045aae5a7bb37c01a635a40ce6f54/test/EFCore.SqlServer.FunctionalTests/SqlServerDatabaseCreatorTest.cs#L100-L155):
* Does it fail only for the above combo of args? (`async: False, open: True, ambientTransaction: False`)
* Does it fail if you run just [one of the test cases](https://github.com/aspnet/EntityFrameworkCore/blob/f6bb8c19a37045aae5a7bb37c01a635a40ce6f54/test/EFCore.SqlServer.FunctionalTests/SqlServerDatabaseCreatorTest.cs#L111-L114) in parallel?
* Would it be possible to just do the same in a standalone console app (compile in `BloggingContext` and `SqlServerTestStore`)? ... That would significantly simplify debugging.
* Are there any parallel writes to the underlying SslStream? Maybe from `DbConnectionPool` usage, caused by the test running 4x in parallel? -- The information if it fails only when running in parallel vs. in  a loop will be valuable IMO.

**Comment by saurabh500** [comment]
Sounds good. This repro has a lot of layers and done of which are out of my area of expertise. If EF detects this failure in their CI they could ping on this bug.
 I will close this bug for now until more info comes from upstream (in case  EF experiences failures). I was checking the state of SqlClient with EF tests and opened this bug before we get such bugs in ask mode.

**Comment by bgrainger** [comment]
This same error has been reported from users of MySqlConnector at mysql-net/MySqlConnector#509:

```
System.NotSupportedException:  The WriteAsync method cannot be called when another write operation is pending.
   at System.Net.Security.SslStreamInternal.WriteAsyncInternal[TWriteAdapter](TWriteAdapter writeAdapter, ReadOnlyMemory`1 buffer)
   at System.Net.Security.SslStreamInternal.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Security.SslStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at MySqlConnector.Protocol.Serialization.StreamByteHandler.WriteBytesAsync(ArraySegment`1 data, IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Protocol\Serialization\StreamByteHandler.cs:line 87
--- End of stack trace from previous location where exception was thrown ---
   at MySqlConnector.Utilities.ValueTaskExtensions.ContinueWith[T,TResult](ValueTask`1 valueTask, Func`2 continuation) in C:\projects\mysqlconnector\src\MySqlConnector\Utilities\ValueTaskExtensions.cs:line 8
   at MySqlConnector.Core.ServerSession.DisposeAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ServerSession.cs:line 200
   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ConnectionPool.cs:line 89
   at MySqlConnector.Core.ConnectionPool.GetSessionAsync(MySqlConnection connection, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ConnectionPool.cs:line 141
   at MySql.Data.MySqlClient.MySqlConnection.CreateSessionAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlConnection.cs:line 399
   at MySql.Data.MySqlClient.MySqlConnection.OpenAsync(Nullable`1 ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlConnection.cs:line 170
   at MySql.Data.MySqlClient.MySqlConnection.Open() in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlConnection.cs:line 156
   at BackOffice.Controllers.ApiController.GetCustomerList(Nullable`1 id) in C:\Dev\New Line Mobile\NewLineBilling\BackOffice\Controllers\ApiController.cs:line 43
   at lambda_method(Closure , Object , Object[] )
```

It's also been reported by someone using EF.Core with Pomelo.EntityFrameworkCore.MySql (the call stack above is just ADO.NET).

Given that it has repro'ed both with and without EF.Core, and with both MySqlConnector and SqlClient, and the users reported it started happening after updating to .NET Core 2.1, this strongly suggests (to me) that a change in `SslStream` is responsible. (Although, as you've stated, it _may_ be a change that's exposing a bug in higher-level code.)

One interesting thing to note is that both the MySqlConnector and SqlConnection call stacks show a call to `Dispose` on a higher-level object (`System.Data.SqlClient.SNI.SNIMarsHandle.Dispose`, `MySqlConnector.Core.ServerSession.DisposeAsync`) which is (I assume) sending some kind of "Close" packet over the `SslStream` (presumably before disposing it).

I can't tell if this means that the higher-level code is in some sort of error-recovery or inconsistent state and incorrectly using the `SslStream` (and this same bug just coincidentally happens to be in both MySqlConnector and SqlClient), or if some invariant in `SslStream` has changed, or what.

Unfortunately it repros very infrequently (I've never personally seen it) so I don't have a test case. @stephentoub @karelz Could someone who's familiar with the SslStream changes provide some informed speculation about what might be going on here?

**Comment by bgrainger** [comment]
Is it possible an exception type changed; e.g., this code might have been throwing `ObjectDisposedException` before, but is now throwing `NotSupportedException`?

**Comment by Spaceman1861** [comment]
I believe we are getting the same error:

```
System.NotSupportedException:  The WriteAsync method cannot be called when another write operation is pending.
   at System.Net.Security.SslStreamInternal.WriteAsyncInternal[TWriteAdapter](TWriteAdapter writeAdapter, ReadOnlyMemory`1 buffer)
   at System.Net.Security.SslStreamInternal.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.Net.Security.SslStream.Write(Byte[] buffer, Int32 offset, Int32 count)
   at System.Data.SqlClient.SNI.SNITCPHandle.Send(SNIPacket packet)
   at System.Data.SqlClient.SNI.SNIMarsConnection.Send(SNIPacket packet)
   at System.Data.SqlClient.SNI.SNIMarsHandle.SendControlPacket(SNISMUXFlags flags)
   at System.Data.SqlClient.SNI.SNIMarsHandle.Dispose()
   at System.Data.SqlClient.SNI.TdsParserStateObjectManaged.Dispose()
   at System.Data.SqlClient.TdsParser.Disconnect()
   at System.Data.SqlClient.SqlInternalConnectionTds.Dispose()
   at System.Data.ProviderBase.DbConnectionPool.DestroyObject(DbConnectionInternal obj)
   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, UInt32 waitForMultipleObjectsTimeout, Boolean allowCreate, Boolean onlyOneCheckConnection, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionPool.TryGetConnection(DbConnection owningObject, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionFactory.TryGetConnection(DbConnection owningConnection, TaskCompletionSource`1 retry, DbConnectionOptions userOptions, DbConnectionInternal oldConnection, DbConnectionInternal& connection)
   at System.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(DbConnection outerConnection, DbConnectionFactory connectionFactory, TaskCompletionSource`1 retry, DbConnectionOptions userOptions)
   at System.Data.SqlClient.SqlConnection.TryOpen(TaskCompletionSource`1 retry)
   at System.Data.SqlClient.SqlConnection.OpenAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnection.OpenAsync(CancellationToken cancellationToken, Boolean errorsExpected)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.BufferlessMoveNext(Boolean buffer, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Storage.ExecutionStrategy.ExecuteImplementationAsync[TState,TResult](Func`4 operation, Func`4 verifySucceeded, TState state, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable`1.AsyncEnumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.FirstOrDefault_[TSource](IAsyncEnumerable`1 source, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.TaskResultAsyncEnumerable`1.Enumerator.MoveNext(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.SelectEnumerableAsyncIterator`2.MoveNextCore(CancellationToken cancellationToken)
   at System.Linq.AsyncEnumerable.AsyncIterator`1.MoveNext(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.AsyncLinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteSingletonAsyncQuery[TResult](QueryContext queryContext, Func`2 compiledQuery, IDiagnosticsLogger`1 logger, Type contextType)
```

We are running dotnetcore 2.1.2 and efcore 2.0.1.

This is occurring in a simple read operation.

Just adding my stack trace if it helps.

If anyone needs any particular info let me know.

Are there any known work arounds for this?

**Comment by karelz** [comment]
The best chance to make it actionable is for someone with a repro to try to make a general repro out of it (ideally minimal) and pass it to us.
Or try to trace it with logging, etc.

Is anyone in that position?

**Comment by fuhry** [comment]
Hey, I know this is an old issue, but we're seeing this exact same trace in a Bedrock framework based server application we are developing internally at @ZorusTech, and it doesn't look like a resolution was ever reached here.

We're working on a good minimal reproduction case and will comment on this issue with a link to the repo once it's up.

**Comment by karelz** [comment]
Please file a new issue and link this one if you get something actionable.
